web_api.md

```diff
@@ -0,0 +1,337 @@
+      and workspace ID.
+    * In both cases, a job ID is returned.
+5. With the given job ID, it is possible to check the job status by calling:
+    * `GET /processor/{executable}/{job-id}` for a single processor, or
+    * `GET /workflow/{workflow-id}/{job-id}` for the workflow.
+6. Download the resulting workspace via the `GET /workspace/{workspace-id}` endpoint and get back an OCRD-ZIP.
+   Set the request header to `Accept: application/json` in case you only want the meta-data of the workspace but not the
+   files.
+
+## Suggested OCR-D System Architecture
+
+There are various ways to build a system which implements this Web API. In this section, we describe a distributed
+architecture, which greatly improves the scalability, flexibility, and reliability of the system compared to
+the [CLI](https://ocr-d.de/en/spec/cli) and the Distributed Processor REST Calls approach.
+
+<figure>
+  <img src="images/web-api-distributed-queue.jpg" alt="Distributed architecture with the Web API"/>
+  <figcaption align="center">
+    <b>Fig. 1:</b> OCR-D System Architecture
+  </figcaption>
+</figure>
+
+In this architecture, all servers are implemented using [FastAPI](https://fastapi.tiangolo.com/). Behind the scene, it
+runs [Uvicorn](https://www.uvicorn.org/), an [ASGI](https://asgi.readthedocs.io/en/latest/) web server implementation
+for Python. [RabbitMQ](https://www.rabbitmq.com/) is used for the Process Queue, and [MongoDB](https://www.mongodb.com/)
+is the database system. There are many options for a reverse proxy, such as Nginx, Apache, or HAProxy. From our side, we
+recommend using [Traefik](https://doc.traefik.io/traefik/).
+
+### Description
+
+As shown in Fig. 1, each section in the [Web API specification](#the-specification) is implemented by different servers,
+which are Discovery Server, Processing Server, Workflow Server, and Workspace Server respectively. Although each server
+in the figure is deployed on its own machine, it is completely up to the implementors to decide which machines run which
+servers. However, having each processor run on its own machine reduces the risk of version and resource conflicts.
+Furthermore, the machine can be customized to best fit the processor's hardware requirements and throughput demand. For
+example, some processors need GPU computation, while others do not, or some need more CPU capacity while others need
+more memory.
+
+**Processing**: since the `Processing` section is provided by [OCR-D Core](https://github.com/OCR-D/core), implementors
+do not need to implement Processing Server, Process Queue, and Processing Worker themselves, they can reuse/customize
+the existing implementation. Once a request arrives, it will be (split into single-processor jobs and) pushed to a
+job queue. A job queue always has the same name as its respective processors. For example, `ocrd-olena-binarize`
```

Comment:
```suggestion
do not need to implement Workflow Server, Processing Server, Process Queue, and Processing Worker themselves, they can reuse/customize
the existing implementation. Once a workflow request arrives, it will be split into single-processor jobs
to be scheduled as processing requests whenever the previous job's result comes available. 
Each processing request will have a job pushed to the job queue named after the respective processor executable. For example, `ocrd-olena-binarize`
```

Reply:
I would not mention the Workflow Server here, since we're talking about the `Processing` part of the API in this paragraph. And with the opening sentence:

> since the `Processing` section is provided by [OCR-D Core](https://github.com/OCR-D/core),...

adding the Workflow Server here doesn't fit.

Reply:
Yes, I can see it more clearly now. But then the parenthesis (which from my suggestion) must be removed here, too. (Nothing needs to be split up, requests are already atomic. Except if splitting should happen here from document to page level...)

Perhaps the (split/monitor) relation between workflow requests and processing requests can be explained by a dedicated `Workflow` section then? (You do have the Workflow Server in the architecture figure and in the glossary...)

Reply:
Agreed!